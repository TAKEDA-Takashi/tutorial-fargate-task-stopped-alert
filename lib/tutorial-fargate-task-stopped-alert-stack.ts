import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as targets from '@aws-cdk/aws-events-targets';

import { ScheduledFargateTask } from '@aws-cdk/aws-ecs-patterns';
import { Schedule } from '@aws-cdk/aws-applicationautoscaling';
import { Topic } from '@aws-cdk/aws-sns';
import { EmailSubscription } from '@aws-cdk/aws-sns-subscriptions';
import { Rule } from '@aws-cdk/aws-events';

export class TutorialFargateTaskStoppedAlertStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'Vpc', {
      cidr: '10.0.0.0/16',
      maxAzs: 1,
    });

    const cluster = new ecs.Cluster(this, 'Cluster', {
      vpc,
    });

    const task = new ScheduledFargateTask(this, 'ScheduledFargateTask', {
      cluster,
      schedule: Schedule.expression('cron(0/5 * * * ? *)'),
      scheduledFargateTaskImageOptions: {
        memoryLimitMiB: 512,
        cpu: 256,
        command: [
          'sh',
          '-c',
          'sleep 5',
        ],
        image: ecs.ContainerImage.fromRegistry('amazonlinux:2'),
      },
    });

    const taskStoppedAlert = new Topic(this, 'TaskStoppedAlert', {
      topicName: 'TaskStoppedAlert',
    });

    taskStoppedAlert.addSubscription(new EmailSubscription('alert@example.com'));

    new Rule(this, 'Rule', {
      eventPattern: {
        source: ['aws.ecs'],
        detailType: ['ECS Task State Change'],
        detail: {
          lastStatus: ['STOPPED'],
          stoppedReason: ['Essential container in task exited'],
        },
      },
      targets: [new targets.SnsTopic(taskStoppedAlert)],
    });
  }
}
