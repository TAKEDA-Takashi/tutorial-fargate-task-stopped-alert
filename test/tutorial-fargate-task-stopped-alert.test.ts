import * as cdk from '@aws-cdk/core';
import { SynthUtils } from '@aws-cdk/assert';
import { TutorialFargateTaskStoppedAlertStack } from '../lib/tutorial-fargate-task-stopped-alert-stack';

test('Stack Snapshot test', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new TutorialFargateTaskStoppedAlertStack(app, 'MyTestStack', {
    env: {
      account: '123456789012',
      region: 'ap-northeast-1',
    },
  });
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});