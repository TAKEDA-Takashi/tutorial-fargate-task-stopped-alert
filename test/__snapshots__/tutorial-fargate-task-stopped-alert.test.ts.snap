// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Snapshot test 1`] = `
Object {
  "Resources": Object {
    "ClusterEB0386A7": Object {
      "Type": "AWS::ECS::Cluster",
    },
    "Rule4C995B7F": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "lastStatus": Array [
              "STOPPED",
            ],
            "stoppedReason": Array [
              "Essential container in task exited",
            ],
          },
          "detail-type": Array [
            "ECS Task State Change",
          ],
          "source": Array [
            "aws.ecs",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "TaskStoppedAlertF3947265",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduledFargateTaskScheduledEventRule2B79E34F": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0/5 * * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "ClusterEB0386A7",
                "Arn",
              ],
            },
            "EcsParameters": Object {
              "LaunchType": "FARGATE",
              "NetworkConfiguration": Object {
                "AwsVpcConfiguration": Object {
                  "AssignPublicIp": "DISABLED",
                  "SecurityGroups": Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "ScheduledFargateTaskScheduledTaskDefSecurityGroupE075BC19",
                        "GroupId",
                      ],
                    },
                  ],
                  "Subnets": Array [
                    Object {
                      "Ref": "VpcPrivateSubnet1Subnet536B997A",
                    },
                  ],
                },
              },
              "TaskCount": 1,
              "TaskDefinitionArn": Object {
                "Ref": "ScheduledFargateTaskScheduledTaskDef521FA675",
              },
            },
            "Id": "Target0",
            "Input": "{}",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "ScheduledFargateTaskScheduledTaskDefEventsRole6CE19522",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduledFargateTaskScheduledTaskDef521FA675": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Command": Array [
              "sh",
              "-c",
              "sleep 5",
            ],
            "Essential": true,
            "Image": "amazonlinux:2",
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "ScheduledFargateTaskScheduledTaskDefScheduledContainerLogGroup4134B16C",
                },
                "awslogs-region": "ap-northeast-1",
                "awslogs-stream-prefix": "ScheduledFargateTask",
              },
            },
            "Name": "ScheduledContainer",
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "ScheduledFargateTaskScheduledTaskDefExecutionRoleD37356D5",
            "Arn",
          ],
        },
        "Family": "MyTestStackScheduledFargateTaskScheduledTaskDef796E8348",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "ScheduledFargateTaskScheduledTaskDefTaskRoleD0FF16AD",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "ScheduledFargateTaskScheduledTaskDefEventsRole6CE19522": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledFargateTaskScheduledTaskDefEventsRoleDefaultPolicy4903ED72": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ecs:RunTask",
              "Condition": Object {
                "ArnEquals": Object {
                  "ecs:cluster": Object {
                    "Fn::GetAtt": Array [
                      "ClusterEB0386A7",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "ScheduledFargateTaskScheduledTaskDef521FA675",
              },
            },
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduledFargateTaskScheduledTaskDefExecutionRoleD37356D5",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduledFargateTaskScheduledTaskDefTaskRoleD0FF16AD",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledFargateTaskScheduledTaskDefEventsRoleDefaultPolicy4903ED72",
        "Roles": Array [
          Object {
            "Ref": "ScheduledFargateTaskScheduledTaskDefEventsRole6CE19522",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduledFargateTaskScheduledTaskDefExecutionRoleD37356D5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledFargateTaskScheduledTaskDefExecutionRoleDefaultPolicy3E3AEE49": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduledFargateTaskScheduledTaskDefScheduledContainerLogGroup4134B16C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledFargateTaskScheduledTaskDefExecutionRoleDefaultPolicy3E3AEE49",
        "Roles": Array [
          Object {
            "Ref": "ScheduledFargateTaskScheduledTaskDefExecutionRoleD37356D5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduledFargateTaskScheduledTaskDefScheduledContainerLogGroup4134B16C": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "ScheduledFargateTaskScheduledTaskDefSecurityGroupE075BC19": Object {
      "Properties": Object {
        "GroupDescription": "MyTestStack/ScheduledFargateTask/ScheduledTaskDef/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ScheduledFargateTaskScheduledTaskDefTaskRoleD0FF16AD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskStoppedAlertF3947265": Object {
      "Properties": Object {
        "TopicName": "TaskStoppedAlert",
      },
      "Type": "AWS::SNS::Topic",
    },
    "TaskStoppedAlertPolicyCA24B1B0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "TaskStoppedAlertF3947265",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "TaskStoppedAlertF3947265",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "TaskStoppedAlertalertexamplecom040789D6": Object {
      "Properties": Object {
        "Endpoint": "alert@example.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "TaskStoppedAlertF3947265",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcPrivateSubnet1DefaultRouteBE02A9ED": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VpcPublicSubnet1NATGateway4D7517AA",
        },
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet1RouteTableB2C5B500",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPrivateSubnet1RouteTableAssociation70C59FA6": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet1RouteTableB2C5B500",
        },
        "SubnetId": Object {
          "Ref": "VpcPrivateSubnet1Subnet536B997A",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPrivateSubnet1RouteTableB2C5B500": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPrivateSubnet1Subnet536B997A": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.128.0/17",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/PrivateSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet1DefaultRoute3DA9E72A": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet1EIPD7E02669": Object {
      "Properties": Object {
        "Domain": "vpc",
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcPublicSubnet1NATGateway4D7517AA": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VpcPublicSubnet1EIPD7E02669",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcPublicSubnet1RouteTable6C95E38E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet1RouteTableAssociation97140677": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet1Subnet5C2D37C4": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/17",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/PublicSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
